src/VkLayer_GF_amber_scoop/src/draw_call_tracker.cc

```diff
@@ -283,4 +296,149 @@ void DrawCallTracker::CreateIndexBufferDeclarations(
+
+  // Create a list of pipeline barriers that should be used to synchronize
+  // the access to index buffer.
+  std::vector<const CmdPipelineBarrier*> pipeline_barriers;
+
+  // Make span for attribute and binding descriptions for safer access.
+  absl::Span<const VkVertexInputAttributeDescription> attribute_descriptions =
+      absl::MakeConstSpan<>(vertex_input_state.pVertexAttributeDescriptions,
+                            vertex_input_state.vertexAttributeDescriptionCount);
+  absl::Span<const VkVertexInputBindingDescription> binding_descriptions =
+      absl::MakeConstSpan<>(vertex_input_state.pVertexBindingDescriptions,
+                            vertex_input_state.vertexBindingDescriptionCount);
+
+  // Go through all attribute descriptions to get the pipeline's vertex buffer
+  // bindings. First copy the whole |VkBuffer| into a host visible memory.
+  // Copied buffers are stored temporarily in |copied_buffers| map to prevent
+  // the same buffer being copied multiple times if it's used in more than one
+  // binding.
+  for (const auto& attribute_description : attribute_descriptions) {
+    // Find description for the binding used in this vertex attribute.
+    const auto* binding_description = std::find_if(
+        binding_descriptions.begin(), binding_descriptions.end(),
+        [&](const VkVertexInputBindingDescription& _binding_description) {
+          return _binding_description.binding == attribute_description.binding;
+        });
+
+    // Verify that a binding description exists.
+    if (binding_description == nullptr) {
+      LOG("Unable to find |VkVertexInputBindingDescription| for binding [%i] "
+          "in |VkVertexInputAttributeDescription| at location [%i]",
+          attribute_description.binding, attribute_description.location);
+      RUNTIME_ASSERT(false);
+    }
+
+    // Get Vulkan handle and create info for the buffer.
+    const VertexBufferBinding& vertex_buffer =
+        draw_call_state_.bound_vertex_buffers.at(attribute_description.binding);
+    const VkBufferCreateInfo& buffer_create_info =
+        device_data->buffers.Get(vertex_buffer.buffer)->GetCreateInfo();
+
+    // Usage bits should contain |VK_BUFFER_USAGE_VERTEX_BUFFER_BIT|
+    DEBUG_ASSERT(
+        (buffer_create_info.usage & VK_BUFFER_USAGE_VERTEX_BUFFER_BIT) != 0);
+
+    // Check if the buffer has been copied already, i.e. it is stored in the
+    // |copied_buffers| map.
+    if (copied_buffers[vertex_buffer.buffer].get() == nullptr) {
+      std::unique_ptr<BufferCopy> vertex_buffer_copy =
+          std::make_unique<BufferCopy>(
+              device_data, vertex_buffer.buffer, buffer_create_info.size,
+              draw_call_state_.queue, GetCommandPool(device_data));
+
+      copied_buffers[vertex_buffer.buffer] = std::move(vertex_buffer_copy);
+    }
+    const BufferCopy* vertex_buffer_copy =
+        copied_buffers[vertex_buffer.buffer].get();
+
+    std::string buffer_name = "vert_";
```

Comment:
The code below looks really good. But I can't help asking if we can do something simpler: I feel we should perhaps always copy the entire buffer. Then, could we also always define the buffer in AmberScript as UINT8. It looks like we then can't specify the FORMAT and OFFSET when binding the vertex data, but I think we could add those as optional parameters in Amber quite easily. I will mention this on the chat.